### ISSUE TITLE
daemon: look into using semaphores to simplify queuing endpoint builds

### ISSUE THREAD

**Issue by ianvernon** [cilium/cilium#3289]
`daemon/daemon.go:StartEndpointBuilders` creates logic with goroutines to accomplish a capped number of goroutines for rebuilding endpoints. After discussion with @rlenglet , this could be simplified using semaphores with a specified weight. 

I suggest we look into using golang's semaphore package to accomplish this: https://godoc.org/golang.org/x/sync/semaphore

It would make the code easier to understand and reduce the complexity around queueing endpoint builds. I'm open to other opinions on this, of course. Thoughts?

**Comment by aanm** [comment]
1) Does that package allows the endpoint builds to be made in the same order they were requested?

2) Won't this means the requests to perform an endpoint build will be blocked until the semaphore is available?

**Comment by rlenglet** [comment]
@aanm

1. This semaphore implementation guarantees that the calls to `Acquire` return in order. So yes, this will guarantee that the endpoint builds are made in the same order they were requested.
2. Isn't that already the case?

**Comment by ianvernon** [comment]
We could also look into using https://godoc.org/golang.org/x/sync/singleflight to ensure synchronicity for endpoint builds:

```
// Do executes and returns the results of the given function, making
// sure that only one execution is in-flight for a given key at a
// time. If a duplicate comes in, the duplicate caller waits for the
// original to complete and receives the same results.
// The return value shared indicates whether v was given to multiple callers.
```

**Comment by rlenglet** [comment]
I don’t think the singleflight lib would be usable here because our functions have lots of side effects.

**Comment by ianvernon** [comment]
Closing this; buildqueue functionality was merged in #5446 .
